src/algorithms/graphs/breadthFirstSearch.java

```diff
@@ -0,0 +1,133 @@
+package src.algorithms.graphs;
+
+import java.util.*;
+
+import src.algorithms.graphs.util.BinaryTreeNode;
+import src.algorithms.graphs.util.GraphNode;
+
+/**
+ * Implementation of BFS
+ *
+ * Breadth-First search is a graph traversal algorithm that utilizes a queue (FIFO) data structure
+ * It is useful in finding a shortest-hops solution in graphs
+ * This method is also used to obtain level order traversals of trees.
+ *
+ * In general, BFS works as such:
+ *  - Start with a queue that contains the root node
+ *  - While the queue is not empty:
+ *      - Pop a vertex from the queue
+ *      - Push all neighbours to the queue if they have not been visited yet
+ *      - Update any variables as needed
+ *
+ * Time: O(V + E), where V is the number of vertices/nodes, and E is the number of edges in the graph
+ * Explanation: Each vertex is popped in O(1) time from the stack exactly once, hence O(V)
+ *              For each edge, we must check in O(1) time if we have visited the adjacent vertex to know
+ *              whether to push it into the queue, hence O(E).
+ *              Note that if the graph is a forest, BFS will likely terminate earlier, as fewer vertices are traversed
+ *
+ * Space: O(V): We utilize a Hashset to store the vertices we have visited already. In the worst case we have a
+ *              connected graph where all vertices are traversed, and our Hashset stores all O(V) vertices.
+ *              Further, we use a queue to hold the vertices to be traversed. In the worst case, the root will have all
+ *              other vertices as neighbours, and our queue will contain O(V) vertices.
+ *
+ * ** Note: The above description assumes an adjacency list in order to consider neighbours in an efficient manner
+ *          If an adjacency matrix were used, it would cost O(V) to find neighbours for a single vertex, making our
```

Comment:
not just average! Best as well. If we have a connected graph of v nodes, we will surely 'explore' all of them. Since we do not have an efficient way of querying and getting their neighbors instantly, we have no choice but to do O(v) search of all other nodes to check if it is a neighbor of this node.